### PR TITLE
Support for the gRPC Docker registry proxy

### PR DESCRIPTION
**Pull request by mikejholly**
This PR adds a new gRPC service that makes use of the pre-existing gRPC connection to stream registry image data out of the embedded registry and into the local container frontend (`docker pull`, etc.). 

image

Associated with: https://github.com/earthly/earthly/pull/3299

### CODE REVIEW

**Review by vladaionescu — APPROVED** [review]
LGTM! Perhaps @alexcb should also review this.

File: api/services/registry/server.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	s.last = msg.GetData()
+	n := copy(p, s.last)
+	s.last = s.last[n:]
+
+	return n + l, nil
+}
+
+// parseHeader parses the incoming request header and extracts the method, path,
+// & header values.
+func parseHeader(r io.Reader) (method string, path string, header http.Header, err error) {
```
Comment (alexcb) [inline api/services/registry/server.go:75]: so you've created your own webserver I see? interesting :)

(if it works, it works.)

**Review by alexcb — APPROVED** [review]
I think it looks good -- my only concern is if we find ourselves needing to proxy anything else this is tied to the http protocol. I think it's fine for now -- if (and this is a big if) we need to proxy to a different service, i think it would be good to refactor this to be more generic (e.g. `Proxy` could take some sort of serviceID).

**Please rebase all your commits prior to merging this PR**

**Comment by mikejholly** [comment]
> I think it looks good -- my only concern is if we find ourselves needing to proxy anything else this is tied to the http protocol. I think it's fine for now -- if (and this is a big if) we need to proxy to a different service, i think it would be good to refactor this to be more generic (e.g. `Proxy` could take some sort of serviceID).

Interesting point. This service is purpose-built to speak with the embedded registry, but it's certainly possible to register other services on the gRPC connection for other purposes. We could definitely take a look at making this more generic, like you suggested, as the need arises. I may also take a look at streaming the raw HTTP data over a lower-level TCP connection. I figured the HTTP libraries would help with some of the overhead, but that may not be necessary.

**Comment by mikejholly** [comment]
On second thought, I think I may leave this as is for the time being. We can always revisit it later if needed.

**Pull request merged**